File: server/src/models/Plan.ts
import mongoose, { Document, Schema } from 'mongoose';

export interface IPlan extends Document {
    userId: Schema.Types.ObjectId;
    calories: number;
    carbsGrams: number;
    proteinGrams: number;
    fatsGrams: number;
    healthScore: number; // 0..10
    targetChangeLbs?: number; // e.g., -11 for lose 11 lbs
    targetDateIso?: string; // ISO date string
    maintenanceCalories?: number; // Daily maintenance calories
    calorieDeficit?: number; // Positive for weight loss, negative for weight gain
    dailyGoal?: string; // "lose_weight", "gain_weight", or "maintain_weight"
    createdAt: Date;
    updatedAt: Date;
}

const planSchema = new Schema<IPlan>(
    {
        userId: { type: Schema.Types.ObjectId, ref: 'User', index: true, required: true },
        calories: { type: Number, required: true },
        carbsGrams: { type: Number, required: true },
        proteinGrams: { type: Number, required: true },
        fatsGrams: { type: Number, required: true },
        healthScore: { type: Number, required: true },
        targetChangeLbs: { type: Number, required: false },
        targetDateIso: { type: String, required: false },
        maintenanceCalories: { type: Number, required: false },
        calorieDeficit: { type: Number, required: false },
        dailyGoal: { type: String, required: false },
    },
    { timestamps: true }
);

const Plan = mongoose.model<IPlan>('Plan', planSchema);

export default Plan;

File: lib/features/plan/domain/entities/plan.dart
class PlanEntity {
  final int calories;
  final int carbsGrams;
  final int proteinGrams;
  final int fatsGrams;
  final int healthScore; // 0..10
  final double? targetChangeLbs; // negative = lose, positive = gain
  final DateTime? targetDate;
  final int? maintenanceCalories;
  final int? calorieDeficit; // positive for loss, negative for gain
  final String? dailyGoal; // lose_weight | gain_weight | maintain_weight

  const PlanEntity({
    required this.calories,
    required this.carbsGrams,
    required this.proteinGrams,
    required this.fatsGrams,
    required this.healthScore,
    this.targetChangeLbs,
    this.targetDate,
    this.maintenanceCalories,
    this.calorieDeficit,
    this.dailyGoal,
  });

  factory PlanEntity.fromJson(Map<String, dynamic> json) {
    final plan = json['plan'] is Map<String, dynamic>
        ? json['plan'] as Map<String, dynamic>
        : json;
    return PlanEntity(
      calories: (plan['calories'] as num).toInt(),
      carbsGrams: (plan['carbsGrams'] as num).toInt(),
      proteinGrams: (plan['proteinGrams'] as num).toInt(),
      fatsGrams: (plan['fatsGrams'] as num).toInt(),
      healthScore: (plan['healthScore'] as num).toInt(),
      targetChangeLbs: plan['targetChangeLbs'] != null
          ? (plan['targetChangeLbs'] as num).toDouble()
          : null,
      targetDate: plan['targetDateIso'] != null
          ? DateTime.tryParse(plan['targetDateIso'] as String)
          : null,
      maintenanceCalories: plan['maintenanceCalories'] == null
          ? null
          : (plan['maintenanceCalories'] as num).toInt(),
      calorieDeficit: plan['calorieDeficit'] == null
          ? null
          : (plan['calorieDeficit'] as num).toInt(),
      dailyGoal: plan['dailyGoal'] as String?,
    );
  }
}

